### ISSUE TITLE
Hash alg for PCR selection not available

### ISSUE THREAD

**Issue by kwmartin** [tpm2-software/tpm2-tools#3215]
New to TPM2 and trying to get anything working (with poor success).
> tss2_provision
ERROR:fapi:src/tss2-fapi/ifapi_helpers.c:2276:ifapi_check_profile_pcr_selection() Hash alg for PCR selection not available. ErrorCode (0x0006000b) 
ERROR:fapi:src/tss2-fapi/api/Fapi_Provision.c:555:Fapi_Provision_Finish() Invalid PCR selection in profile. ErrorCode (0x0006000b) 
ERROR:fapi:src/tss2-fapi/api/Fapi_Provision.c:174:Fapi_Provision() ErrorCode (0x0006000b) Provision 
Fapi_Provision(0x6000B) - fapi:A parameter has a bad value
> 
Part of the file /etc/tpm2-tss/fapi-profiles/P_ECCP256SHA256.json has:
    "pcr_selection": [
       { "hash": "TPM2_ALG_SHA1",
         "pcrSelect": [ ],
       },
       { "hash": "TPM2_ALG_SHA256",
         "pcrSelect": [ 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23 ]
       }
    ],
I'm not sure where I look to see if hash algorithms are installed. Also, doing a tpm2_pcrread seems to indicate SHA256 is installed in the TPM

> tpm2_pcrread
  sha256:
    0 : 0xC7BF5F0ABA3E1C1F77E8F5B1FC6506DD7DF21D5E5AD12FADEC19256059B16F17
    1 : 0x16BB306F21F1D8141F0568E63968F438771A3927DDCA0BC4BECA68DF5FB80042
    2 : 0xEE7295E4527C7216E3F73D905BFCF6C2568CB992780E4A99B8EC79BC686153E9
    3 : 0x3D458CFE55CC03EA1F443F1562BEEC8DF51C75E14A9FCF9A7234A13F198E7969
    4 : 0xB50B80F287AD601C34C5E86C228FAF2228E27FF319B07910B6B86FBE63872C28
    5 : 0x57C444BDA294A2763BB3F19A83E9E705176BFD35E457C4D34C5B4E68E5B87893
    6 : 0xBB1A08D0769CAB7973754E1417551E9F8ACE5A3C483D71726AA2F26DD3A25FCB
    7 : 0x3A7C1E1003A3D7016160EE9369191B60C55FDF5FCFE2742C85A02B4679AD5D36
    8 : 0x139F87C2A2848E3BE1E2D86BEE9738BC25981468E0B0AB7687BDC19518DC1649
    9 : 0xB08E0B9AAE3CAA61A14B3447267ABDA6EC5F57BEBAAC3623BC88232F1CEFC168
    10: 0x8B14FF955CDEFF7B8CC037372C5C6415A956453438F3F591B99BE1C3A1831CCF
    11: 0x0000000000000000000000000000000000000000000000000000000000000000
    12: 0x0000000000000000000000000000000000000000000000000000000000000000
    13: 0x0000000000000000000000000000000000000000000000000000000000000000
    14: 0xE3991B7DDD47BE7E92726A832D6874C5349B52B789FA0DB8B558C69FEA29574E
    15: 0x0000000000000000000000000000000000000000000000000000000000000000
    16: 0x0000000000000000000000000000000000000000000000000000000000000000
    17: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
    18: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
    19: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
    20: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
    21: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
    22: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
    23: 0x0000000000000000000000000000000000000000000000000000000000000000

Also,
> openssl engine -t -c tpm2tss
Initializing
(tpm2tss) TPM2-TSS engine for OpenSSL
 [RSA, RAND]
     [ available ]
40377372777F0000:error:1280006A:DSO support routines:dlfcn_bind_func:could not bind to the requested symbol name:../crypto/dso/dso_dlfcn.c:188:symname(EVP_PKEY_base_id): /usr/lib/x86_64-linux-gnu/engines-3/tpm2tss.so: undefined symbol: EVP_PKEY_base_id
40377372777F0000:error:1280006A:DSO support routines:DSO_bind_func:could not bind to the requested symbol name:../crypto/dso/dso_lib.c:176:
> 
I have no idea  how to define EVP_PKEY_base_id?
Also,
> ldd /sbin/cryptsetup | grep libcryptsetup
/sbin/cryptsetup: /lib/x86_64-linux-gnu/libcryptsetup.so.12: version `CRYPTSETUP_2.6' not found (required by /sbin/cryptsetup)
/sbin/cryptsetup: /lib/x86_64-linux-gnu/libcryptsetup.so.12: version `CRYPTSETUP_2.5' not found (required by /sbin/cryptsetup)
	libcryptsetup.so.12 => /lib/x86_64-linux-gnu/libcryptsetup.so.12 (0x00007fbf3d701000)
> 
> ll /lib/x86_64-linux-gnu/ | grep libcrypt
-rw-r--r--   1 root root  264K Dec 17  2021 libcrypt.a
-rw-r--r--   1 root root  8.7M Feb  6 12:57 libcrypto.a
lrwxrwxrwx   1 root root    14 Feb  6 12:57 libcrypto.so -> libcrypto.so.3
-rw-r--r--   1 root root  4.3M Feb  6 12:57 libcrypto.so.3
lrwxrwxrwx   1 root root    23 Aug  4  2022 libcryptsetup.so -> libcryptsetup.so.12.7.0
lrwxrwxrwx   1 root root    23 Aug  4  2022 libcryptsetup.so.12 -> libcryptsetup.so.12.7.0
-rw-r--r--   1 root root  474K Aug  4  2022 libcryptsetup.so.12.7.0
lrwxrwxrwx   1 root root    35 Dec 17  2021 libcrypt.so -> /lib/x86_64-linux-gnu/libcrypt.so.1
lrwxrwxrwx   1 root root    17 Oct 17 20:53 libcrypt.so.1 -> libcrypt.so.1.1.0
-rw-r--r--   1 root root  195K Dec 17  2021 libcrypt.so.1.1.0
Also,
> tss2_getrandom --numBytes 32 -o - | hexdump
ERROR:fapi:src/tss2-fapi/ifapi_keystore.c:515:rel_path_to_abs_path() ErrorCode (0x00060034) FAPI not provisioned for path: P_ECCP256SHA256/HS/SRK. 
ERROR:fapi:src/tss2-fapi/ifapi_keystore.c:582:ifapi_keystore_load_async() ErrorCode (0x00060034) Object P_ECCP256SHA256/HS/SRK not found. 
ERROR:fapi:src/tss2-fapi/fapi_util.c:876:ifapi_load_primary_async() ErrorCode (0x00060034) Could not open: P_ECCP256SHA256/HS/SRK 
ERROR:fapi:src/tss2-fapi/fapi_util.c:1370:ifapi_get_sessions_async() ErrorCode (0x00060034) Load EK 
ERROR:fapi:src/tss2-fapi/api/Fapi_GetRandom.c:167:Fapi_GetRandom_Async() ErrorCode (0x00060034) Create FAPI session 
ERROR:fapi:src/tss2-fapi/api/Fapi_GetRandom.c:87:Fapi_GetRandom() ErrorCode (0x00060034) GetRandom 
Fapi_GetRandom(0x60034) - fapi:Provisioning was not executed.
> 
This is probably because the first step of tss2_provision didn't work?
Maybe it's not rocket science but there are so many things that can go wrong, it's very difficult to use for even the simplest of applications? And very difficult to figure out how to correct errors? Any help and/or suggestions are appreciated. Thank you.

**Comment by kwmartin** [comment]
Update 2: found https://superuser.com/questions/1484674/tpm-2-0-hardware-software-issue
so I:
> cd /sys/class/tpm/tpm0/ppi
> sudo -E su
> printf 5 > request
and rebooted; I have no idea what this does. At reboot I was asked if I wanted to clear TPM and replied affirmatively by typing F12. After rebooting, I could now run tpm2_clear with no response but I still can not run tss2_provision
> tss2_provision
ERROR:fapi:src/tss2-fapi/ifapi_helpers.c:2276:ifapi_check_profile_pcr_selection() Hash alg for PCR selection not available. ErrorCode (0x0006000b) 
ERROR:fapi:src/tss2-fapi/api/Fapi_Provision.c:555:Fapi_Provision_Finish() Invalid PCR selection in profile. ErrorCode (0x0006000b) 
ERROR:fapi:src/tss2-fapi/api/Fapi_Provision.c:174:Fapi_Provision() ErrorCode (0x0006000b) Provision 
Fapi_Provision(0x6000B) - fapi:A parameter has a bad value
> 
My configuration files (shown below) are as they came; so once again stuck at basically the very beginning. I probably shouldn't wine, is this ever difficult!
> c /etc/tpm2-tss/fapi-config.json
{
     "profile_name": "P_ECCP256SHA256",
     "profile_dir": "/etc/tpm2-tss/fapi-profiles/",
     "user_dir": "~/.local/share/tpm2-tss/user/keystore",
     "system_dir": "/var/lib/tpm2-tss/system/keystore",
     "tcti": "",
     "system_pcrs" : [],
     "log_dir" : "/run/tpm2-tss/eventlog/",
     "firmware_log_file": "/sys/kernel/security/tpm0/binary_bios_measurements",
     "ima_log_file": "/sys/kernel/security/ima/binary_runtime_measurements"
}
> 
> c /etc/tpm2-tss/fapi-profiles/P_ECCP256SHA256.json
{
    "type": "TPM2_ALG_ECC",
    "nameAlg":"TPM2_ALG_SHA256",
    "srk_template": "system,restricted,decrypt,0x81000001",
    "srk_description": "Storage root key SRK",
    "srk_persistent": 0,
    "ek_template":  "system,restricted,decrypt",
    "ek_description": "Endorsement key EK",
    "ecc_signing_scheme": {
        "scheme":"TPM2_ALG_ECDSA",
        "details":{
            "hashAlg":"TPM2_ALG_SHA256"
        },
    },
    "sym_mode":"TPM2_ALG_CFB",
    "sym_parameters": {
        "algorithm":"TPM2_ALG_AES",
        "keyBits":"128",
        "mode":"TPM2_ALG_CFB"
    },
    "sym_block_size": 16,
    "pcr_selection": [
       { "hash": "TPM2_ALG_SHA1",
         "pcrSelect": [ ],
       },
       { "hash": "TPM2_ALG_SHA256",
         "pcrSelect": [ 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23 ]
       }
    ],
    "curveID": "TPM2_ECC_NIST_P256",
    "ek_policy": {
        "description": "Endorsement hierarchy used for policy secret.",
        "policy":[
            {
                "type":"POLICYSECRET",
                "objectName": "4000000b",
            }
        ]
    }
}
> 
If anyone can make suggestions on how to proceed, it would be much appreciated.

**Comment by JuergenReppSIT** [comment]
@kwmartin  with `tpm2_getcap algorithms| grep sha -A 9` you can check the available hashalgs and `tpm2_getcap pcrs`
displays the pcrs.
To prevent comming into the lockout mode you can use the following command:
`tpm2_dictionarylockout -Tdevice:/dev/tpmrm0 --setup-parameters --max-tries=<your number>  --clear-lockout`
`

**Comment by kwmartin** [comment]
Thanks Juergen, I ran the commands you suggested (results below). However, before your suggestion came in, just blindly trying things, I deleted:
{ "hash": "TPM2_ALG_SHA1",
"pcrSelect": [ ],
},
from /etc/tpm2-tss/fapi-profiles/P_ECCP256SHA256.json
I could then run tpm2_clear and tpm2_sessionconfig and finally get started. I have no idea why this worked. The tpm2_getcap returns are below (but after the above deletion, I am mostly up and running). I also cloned tpm2-pytss to get the 'pytests' and most of them are now running. Thank you for your help appreciated. A very very difficult start up (for me?), but now that I finally have something working, I'm sure I can go ahead. Maybe sometime I should write a how-to for newbies like myself?

> tpm2_getcap algorithms| grep sha -A 9
sha1:
  value:      0x4
  asymmetric: 0
  symmetric:  0
  hash:       1
  object:     0
  reserved:   0x0
  signing:    0
  encrypting: 0
  method:     0
--
sha256:
  value:      0xB
  asymmetric: 0
  symmetric:  0
  hash:       1
  object:     0
  reserved:   0x0
  signing:    0
  encrypting: 0
  method:     0
sha384:
  value:      0xC
  asymmetric: 0
  symmetric:  0
  hash:       1
  object:     0
  reserved:   0x0
  signing:    0
  encrypting: 0
  method:     0
--
sha3_256:
  value:      0x27
  asymmetric: 0
  symmetric:  0
  hash:       1
  object:     0
  reserved:   0x0
  signing:    0
  encrypting: 0
  method:     0
sha3_384:
  value:      0x28
  asymmetric: 0
  symmetric:  0
  hash:       1
  object:     0
  reserved:   0x0
  signing:    0
  encrypting: 0
  method:     0
> 
> tpm2_getcap pcrs
selected-pcrs:
  - sha256: [ 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23 ]
>
